src/core/kd_tree.rs

```diff
@@ -267,22 +302,21 @@ impl <T: Geometry> KdTree<T> {
     }
 }
 
-impl <T: Geometry> Geometry for KdTree<T> {
+impl<T: Geometry> Geometry for KdTree<T> {
     fn bound(&self) -> BoundingBox {
         self.bound.clone()
     }
 
     fn intersect(&self, ray: &Ray) -> Option<Intersection> {
-        intersect(&self, ray.clone())
+        intersect(self, ray.clone()).min_by_key(|i| OrderedFloat(i.distance))
```

Comment:
I'm not sure if we should use `NotNaN` or `OrderedFloat` here. The only difference is that `NotNaN` actually makes sure the argument isn't `NaN` and panics if it is, whereas `OrderedFloat` asserts ordering semantics for `NaN` values. They're both `Ord`. We don't want `NaN` values ever but I guess the sanity-checking is helpful?